### ISSUE TITLE
How to activate, and keep frozen, multiple adapters?

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/peft#1374]
Hello all,

I have been working on multiple adapters and part of my project requires that I activate all the loaded adapters. However, they must be frozen. I am running this code:

```python
adapters_items = iter(tqdm.tqdm(adapters.items()))
first_item = next(adapters_items)
model_peft = PeftModel.from_pretrained(model, first_item[1], first_item[0], is_trainable=False)

for adapter_name, model_id in adapters_items:
    model_peft.load_adapter(model_id, adapter_name, is_trainable=False)

model_peft.base_model.set_adapter(list(adapters.keys()))
```

After some debugging, I see that the adapters are frozen (requires_grad=False) until the last line where I set the active adapters. After they are set to be active, requires_grad=True.

I see that `set_adapter` calls this function on all the LoraLayers, and how it sets the adapters to trainable.
> https://github.com/huggingface/peft/blob/ebbff4023ad276cbcb2466fd7e99be7d3ae0ae11/src/peft/tuners/tuners_utils.py#L464-L484

How can I set the active adapter(s) while keeping them frozen?

**Comment by louieworth** [comment]
I think this [tutorial](https://huggingface.co/docs/peft/main/en/developer_guides/mixed_models#working-with-mixed-adapter-types) can solve this problem:

```python
from peft import PeftMixedModel

base_model = ...  # load the base model, e.g. from transformers
# load first adapter, which will be called "default"
peft_model = PeftMixedModel.from_pretrained(base_model, <path_to_adapter1>)
peft_model.load_adapter(<path_to_adapter2>, adapter_name="other")
peft_model.set_adapter(["default", "other"])
```
The last line is necessary if you want to activate both adapters, otherwise, only the first adapter would be active.

**Comment by EricLBuehler** [comment]
Thanks!

**Comment by BenjaminBossan** [comment]
Wanted to chime in: Only use `PeftMixedModel` if you want to mix _different types of adapters_, e.g. LoRA and LoHa. If all adapters are of the same type (e.g. all LoRA), please use `PeftModel`.

IIUC, the question is about `requires_grad` being `True` after calling `set_adapter`, but you want it to be `False`. What you should do instead is to call `model_peft.required_grad_(False)` to freeze all parameters. Or, if you want to only freeze certain parameters, something like this should work:

```python
model_peft.base_model.set_adapter(list(adapters.keys()))

for name, param in model_peft.named_parameters():
    if <some check on `name`>:  # e.g. `if "lora" in name`
        param.requires_grad = False
```

Note that right now, calling `set_adapter` _always_ activates the gradients for these adapters. I guess what we could do is to check if `is_trainable=False` for the respective configs and not activate the gradients for those. I guess this is somewhat niche, but might lead to fewer surprises in cases like this.

**Comment by EricLBuehler** [comment]
Ok, thank you for the clarification. Could I submit a PR to implement that check?

**Comment by BenjaminBossan** [comment]
> Could I submit a PR to implement that check?

Possibly, but it will unfortunately not be quite trivial. The problem is that `set_adapter` works on individual layers, which don't have a reference to all the configs. So this change would have to be a bit more involved. Another issue could be that if we change the behavior, it is technically backwards incompatible, not sure how big of a problem that would be in practice.

**Comment by EricLBuehler** [comment]
True, programs may be relying on this behavior (although `peft` is technically pre v1). This seems like pretty niche behavior to rely on, and the code currently implements a workaround that would not be negatively affected. It is clearly intentionally implemented this way, and as such would it be possible for me to add an option to one of the configs and submit a PR? This way, it would be an opt-in for now?

**Comment by BenjaminBossan** [comment]
> Alternatively, could we add a separate API, perhaps called `set_adapter_keep_frozen`?

Hmm, I feel like this use case is too narrow to warrant extending the API like this.

Maybe as a first step we could just document the fact that `requires_grad` will be set to `True` for the selected adapter(s) and `False` otherwise. We could also add a snippet for how to set all `requires_grad` to `False` if `True` is not desired to the docstring. WDYT?

**Comment by EricLBuehler** [comment]
That sounds like a great idea. Should I add this to all `set_adapter` functions?
